### PR TITLE
ENH: indexing and __getitem__ of dataframe and series accept zerodim integer np.array as int

### LINKED ISSUE #24919: Please support indexing with 0 dim np.array in df.iloc.
#### Code Sample, a copy-pastable example if possible

The code below works correctly.
```python
df=pd.DataFrame([[1,2],[3,4]])
df.iloc[np.array([0])]
```
But the code below raises an error.
```python
df.iloc[np.array(0)]
```
```
TypeError: object of type 'numpy.int64' has no len()
```

### PR DESCRIPTION
**Pull request by tamuhey**
- [ ] closes #24919
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

- Now the `df.iloc`, `df.loc`, `df.__getitem__` accepts zero dim integer numpy  array, as follows:

```python
df.iloc[np.array(0)]
df.loc[np.array(0)]
df[np.array(0)]
```
- I think this change possibly harms the performance of df.iloc

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Just do “key = lib.item_from_zerodim(key)” at the beginning of the method?

File: pandas/tests/indexing/test_iloc.py
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: can you add a result = and compare to the expected value

**Review by jreback — CHANGES_REQUESTED** [review]
small change, can you add a note in 0.24.1, in Bug Fixes / Indexing

**Comment by jreback** [comment]
we *might* have another open issue for something very similar (which this PR could close), if you can do a quick search would be great.

**Comment by jorisvandenbossche** [comment]
This is not really a bug fix, so no need to include in 0.24.1 IMO

**Comment by jorisvandenbossche** [comment]
Yes, it is incorrect, but it has been like that forever. And you could also see it as a currently unsupported argument type. This also fails for getitem and loc (so this PR is actually introducing an inconsistency).

Backporting is IMO for regressions and important bug fixes. This is not one of those.

**Comment by jreback** [comment]
@jorisvandenbossche fine, won't fight you on it.

@tamuhey put the whatsnew in 0.25.0 pls.

File: pandas/tests/indexing/test_iloc.py
Comment (tamuhey) [reply pandas/tests/indexing/test_iloc.py]: Done!

**Comment by tamuhey** [comment]
Similar issue is found #21946.
I will close this issue after the test is passed

**Comment by tamuhey** [comment]
> > This also fails for getitem and loc (so this PR is actually introducing an inconsistency).
> 
> I think we should actually also consider properly supporting it, if we do down this path.

@jorisvandenbossche
I added the code and `getitem` and `loc` support it.

Should I rename the title of this PR?

**Comment by jorisvandenbossche** [comment]
Yes, changing the title to reflect the changed scope is fine. 
Can you then add a whatsnew notice to v0.25.0.txt ?

**Comment by tamuhey** [comment]
Changed the title!

> Can you then add a whatsnew notice to v0.25.0.txt ?

Should I modify "doc/source/whatsnew/v0.25.0.rst",or add new txt file to the directory?

**Comment by jorisvandenbossche** [comment]
Ah, no, sorry, I meant the existing `v0.25.0.rst` file!

**Comment by tamuhey** [comment]
I think `pd.Series` should also accept zerodim np.array, but now it raises an error as follows:
```python
sr = pd.Series([1,2,3])
sr.iloc[0]                 # this returns 1
sr.iloc[np.array(0)]  # but this raises an error
```
np.ndarray accepts zerodim array for indexing as follows:
```python
arr = np.array([1,2,3])
arr[np.array(0)] # this returns 1
```

I think what I should do is one of the following four:

1. Push commits to resolve this issue
2. Write codes in another branch and create another PR
3. Create issue at first step, write codes, and create PR
4. Do nothing

Which should I do?

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: This is not exactly what I was after. Let's just do a single line for now.

File: pandas/tests/frame/test_indexing.py
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: this comment is not needed

File: pandas/tests/frame/test_indexing.py
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: I would rather have these added test in pandas/tests/indexing, in a sub-file

File: pandas/tests/indexing/test_iloc.py
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: these comments are not needed (and same below)

File: pandas/tests/indexing/test_loc.py
Comment (jreback) [inline pandas/tests/indexing/test_loc.py]: name as s

File: pandas/tests/series/indexing/test_indexing.py
Comment (jreback) [inline pandas/tests/series/indexing/test_indexing.py]: see comment above

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst
Comment (tamuhey) [reply doc/source/whatsnew/v0.25.0.rst]: Sorry, I cannot understand what you mean.
Please tell me more concretely.

File: pandas/tests/frame/test_indexing.py
Comment (tamuhey) [reply pandas/tests/frame/test_indexing.py]: Done!

File: pandas/tests/frame/test_indexing.py
Comment (tamuhey) [reply pandas/tests/frame/test_indexing.py]: Done!

File: pandas/tests/indexing/test_iloc.py
Comment (tamuhey) [reply pandas/tests/indexing/test_iloc.py]: Done!

File: pandas/tests/indexing/test_loc.py
Comment (tamuhey) [reply pandas/tests/indexing/test_loc.py]: Done!

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: on 2nd thought, just add a single line whatsnew entry in Other Enhancements

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst
Comment (tamuhey) [reply doc/source/whatsnew/v0.25.0.rst]: OK!

**Comment by tamuhey** [comment]
@jreback Is it ok this modification?

**Comment by tamuhey** [comment]
@jbrockmendel @jreback @jorisvandenbossche @ryanorsinger Sorry, can anyone tell me what should I do next?

**Review by jreback — COMMENTED** [review]
can u merge master and ping on green

**Comment by jreback** [comment]
will have to wait until i merge #25352 to merge master

**Comment by tamuhey** [comment]
merged! Is it ok?

File: doc/source/whatsnew/v0.25.0.rst
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: use double backticks on DataFrame, Series, np.ndarray., say

you don't need to say ``__getitem__`` as Indexing is pretty general here.

**Review by jreback — CHANGES_REQUESTED** [review]
whatsnew comment, otherwise lgtm. ping on green.

**Comment by tamuhey** [comment]
Done!

**Comment by jorisvandenbossche** [comment]
@tamuhey Thanks!

**Pull request merged**
